src/histogram.c

```diff
@@ -92,108 +389,208 @@ Datum
-	bytea	   *state2 = (PG_ARGISNULL(1) ? NULL : PG_GETARG_BYTEA_P(1));
-	bytea	   *result;
-
+	/* cannot be called directly because of internal-type argument */
 	if (!AggCheckCallContext(fcinfo, &aggcontext))
-	{
-		/* cannot be called directly because of internal-type argument */
 		elog(ERROR, "hist_combinefunc called in non-aggregate context");
-	}
 
+	old = MemoryContextSwitchTo(aggcontext);
 	if (state2 == NULL)
-	{
-		result = copy_state(aggcontext, state1);
-	}
-
+		result = state1;
 	else if (state1 == NULL)
-	{
-		result = copy_state(aggcontext, state2);
-	}
-
+		result = state2;
 	else
 	{
-		Size		i;
 		Datum	   *hist,
 				   *hist_other;
+		Size		i;
 
 		result = copy_state(aggcontext, state1);
-		hist = (Datum *) VARDATA(result);
-		hist_other = (Datum *) VARDATA(state2);
-
-		/* Combine values from state1 and state2 when both states are non-null */
-		for (i = 0; i < HIST_LEN(state1); i++)
-			hist[i] = Int32GetDatum(DatumGetInt32(hist[i]) + DatumGetInt32(hist_other[i]));
+		hist = (Datum *) VARDATA(result->values);
+		hist_other = (Datum *) VARDATA(state2->values);
+		for (i = 0; i < HIST_LEN(result->values); i++)
+			hist[i] = apply_subagg_combine(state1->subagg_state.combinefunc,
+										   fcinfo->context,
+										   hist[i],
+										   hist_other[i]);
 	}
 
-	PG_RETURN_BYTEA_P(result);
+	MemoryContextSwitchTo(old);
+	PG_RETURN_POINTER(result);
 }
 
 /* hist_serializefunc(internal) => bytea */
 Datum
 hist_serializefunc(PG_FUNCTION_ARGS)
 {
-	bytea	   *state;
-	Datum	   *hist;
-	Size		i;
+	HistogramState *state;
+	StringInfoData buf;
+	Datum	   *vals;
+	int			i;
+	HistogramSerializeState *sstate =
+	(HistogramSerializeState *) fcinfo->flinfo->fn_extra;
+
+	if (NULL == sstate)
+	{
+		fcinfo->flinfo->fn_extra = MemoryContextAllocZero(fcinfo->flinfo->fn_mcxt, sizeof(HistogramSerializeState));
+		sstate = (HistogramSerializeState *) fcinfo->flinfo->fn_extra;
+	}
+
+	if (!AggCheckCallContext(fcinfo, NULL))
+		elog(ERROR, "hist_serializefunc function called in non-aggregate context");
 
 	Assert(!PG_ARGISNULL(0));
-	state = PG_GETARG_BYTEA_P(0);
-	hist = (Datum *) VARDATA(state);
+	state = (HistogramState *) PG_GETARG_POINTER(0);
+
+	vals = (Datum *) VARDATA(state->values);
+	pq_begintypsend(&buf);
+	pq_sendint(&buf, HIST_LEN(state->values), 4);
+	pq_sendint(&buf, state->subagg_state.result_type, sizeof(Oid));
+	pq_sendint(&buf, state->subagg_state.sfunc, sizeof(Oid));
+	pq_sendint(&buf, state->subagg_state.finalfunc, sizeof(Oid));
+	pq_sendint(&buf, state->subagg_state.combinefunc, sizeof(Oid));
+	pq_sendint(&buf, state->subagg_state.serialfunc, sizeof(Oid));
+	pq_sendint(&buf, state->subagg_state.deserialfunc, sizeof(Oid));
+
+	/* no serialfn given, get info to serialize each bucket based on type */
+	if (!OidIsValid(state->subagg_state.serialfunc) && !OidIsValid(sstate->fn))
+	{
+		bool		is_varlen;
+
+		getTypeBinaryOutputInfo(state->subagg_state.result_type, &sstate->fn,
```

Comment:
Should we check the result of `is_varlen`? Does it matter?

Reply:
I did steal this from bookends code but I don't believe it matters.

Reply:
Can we verify that it doesn't matter?

Reply:
It doesn't matter since it is used to tell whether the return type from the serialization function is fixed or variable length, but we get the size of the returned bytes later using the `VARSIZE` macro. So whether it is fixed or variable doesn't matter for our purposes, we get the size anyway.